service/oauth2/data_plane.go

```diff
@@ -0,0 +1,50 @@
+package oauth2
+
+import (
+	"strings"
+
+	"golang.org/x/oauth2"
+)
+
+type DataPlaneHelper struct {
+	infos  map[string]*DataPlaneInfo
+	tokens map[string]*oauth2.Token
+}
+
+func (o *DataPlaneHelper) GetDataPlaneDetails(method string, params []string, refresh func(*DataPlaneInfo) (*oauth2.Token, error), infoGetter func() (*DataPlaneInfo, error)) (string, *oauth2.Token, error) {
+	if o.infos == nil {
+		o.infos = make(map[string]*DataPlaneInfo)
+	}
+	if o.tokens == nil {
+		o.tokens = make(map[string]*oauth2.Token)
+	}
+	key := o.generateKey(method, params)
+	info, infoOk := o.infos[key]
+	token, tokenOk := o.tokens[key]
+	if infoOk && tokenOk && token.Valid() {
+		return info.EndpointUrl, token, nil
+	}
```

Comment:
I think we would obtain the same behavior if we were to remove that if right? Assuming that this is correct, doing so will allow us to move the variable instantiation closer to where they are used:

```go
	info, infoOK := o.infos[key]
	if !infoOK {
		// do things ...
	}

	token, tokenOK := o.tokens[key]
	if !tokenOK || !token.Valid() {
		// do other things ...
	}
```

This also have the advantage of better following the Go practice of keeping the "normal flow" at the lowest level of indentation (see https://go.dev/wiki/CodeReviewComments#indent-error-flow).

Alternatively, you could use a `switch` statement to make it clear that you have three distinct code paths:

```go
	token, tokenOK := o.tokens[key]
	info, infoOK := o.infos[key]

	switch {
	case !infoOK:
		// do things ...
	case !tokenOK || !token.Valid():
		// do other things ...
	default:
		// happy path
	}
```

That `switch` statement should produce the same compiled code than what you would obtain with the `ifs`.